File: PathfinderAPI/src/main/java/com/github/ysl3000/bukkit/pathfinding/goals/TalkToStrangers.java
package com.github.ysl3000.bukkit.pathfinding.goals;

import com.github.ysl3000.bukkit.pathfinding.entity.Insentient;
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal;
import org.bukkit.entity.Player;

import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by ysl3000
 */
public class TalkToStrangers implements PathfinderGoal {


    private final int radius;
    private Insentient insentient;

    private List<String> messages;
    private long delay;

    private Iterator<String> currentMessage;
    private long lastTalk = 0;

    public TalkToStrangers(Insentient insentient, List<String> messages, long delay) {
        this.insentient = insentient;
        this.messages = messages;
        this.delay = delay;
        this.radius = 10;
        this.currentMessage = messages.iterator();
    }


    @Override
    public boolean shouldExecute() {
        return isAllowedToTalk() && hasPlayerInRadius() && (currentMessage == null || currentMessage.hasNext());
    }


    @Override
    public boolean shouldContinueUpdating() {
        return currentMessage.hasNext();
    }

    @Override
    public void init() {
        if (currentMessage == null) currentMessage = messages.iterator();
    }

    @Override
    public void execute() {
        if (currentMessage.hasNext() && isAllowedToTalk()) {
            String cMessage = currentMessage.next();
            getNearbyPlayers().forEach(e -> e.sendMessage(cMessage));
            this.lastTalk = System.currentTimeMillis();
        }
    }

    @Override
    public void reset() {
        currentMessage = null;
    }


    private boolean isAllowedToTalk() {
        return (lastTalk + delay) < System.currentTimeMillis();
    }

    private boolean hasPlayerInRadius() {
        return getNearbyPlayers().size() > 0;
    }


    private List<Player> getNearbyPlayers() {
        return insentient.getBukkitEntity().getNearbyEntities(radius, radius, radius).stream().filter(Player.class::isInstance).map(Player.class::cast).collect(Collectors.toList());
    }


}

File: PathfinderAPI/src/main/java/com/github/ysl3000/bukkit/pathfinding/goals/TalkToStrangers.kt
package com.github.ysl3000.bukkit.pathfinding.goals

import com.github.ysl3000.bukkit.pathfinding.entity.Insentient
import com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal
import org.bukkit.entity.Player
import java.util.stream.Collectors

/**
 * Created by ysl3000
 */
class TalkToStrangers(private val insentient: Insentient, private val messages: List<String>, private val delay: Long) : PathfinderGoal {


    private val radius: Int = 10

    private var currentMessage: Iterator<String>? = null
    private var lastTalk: Long = 0


    private val isAllowedToTalk: Boolean
        get() = lastTalk + delay < System.currentTimeMillis()


    private val nearbyPlayers: List<Player>
        get() = insentient.getBukkitEntity().getNearbyEntities(radius.toDouble(), radius.toDouble(), radius.toDouble()).stream()
                .filter { Player::class.java.isInstance(it) }.map<Player> { Player::class.java.cast(it) }.collect(Collectors.toList())

    init {
        this.currentMessage = messages.iterator()
    }


    override fun shouldExecute(): Boolean {
        return isAllowedToTalk && hasPlayerInRadius() && (currentMessage == null || currentMessage!!
                .hasNext())
    }


    override fun shouldTerminate(): Boolean {
        return currentMessage!!.hasNext()
    }

    override fun init() {
        if (currentMessage == null) {
            currentMessage = messages.iterator()
        }
    }

    override fun execute() {
        if (currentMessage!!.hasNext() && isAllowedToTalk) {
            val cMessage = currentMessage!!.next()
            nearbyPlayers.forEach { e -> e.sendMessage(cMessage) }
            this.lastTalk = System.currentTimeMillis()
        }
    }

    override fun reset() {
        currentMessage = null
    }

    private fun hasPlayerInRadius(): Boolean {
        return nearbyPlayers.isNotEmpty()
    }


}
